src/ir/stack-utils.h

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright 2020 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//
+// stack-utils.h: Utilities for manipulating and analyzing stack machine code in
+// the form of Stacky IR.
+//
+// Stacky IR represents stack machine code using normal Binaryen IR types by
+// imposing the following structural constraints:
+//
+//  1. Function bodies and children of control flow (except If conditions) must
+//     be blocks.
+//
+//  2. Blocks may have any Expressions except for Pops as children. The sequence
+//     of instructions in a block follows the same validation rules as in
+//     WebAssembly. That means that any expression may have a concrete type, not
+//     just the final expression in the block.
+//
+//  3. All other children must be Pops, which are used to determine the input
+//     stack type of each instruction. Pops may not have `unreachable` type.
+//
+// For example, the following Binaryen IR Function:
+//
+//   (func $foo (result i32)
+//    (i32.add
+//     (i32.const 42)
+//     (i32.const 5)
+//    )
+//   )
+//
+// would look like this in Stacky IR:
+//
+//   (func $foo (result i32)
+//    (block
+//     (i32.const 42)
+//     (i32.const 5)
+//     (i32.add
+//      (i32.pop)
+//      (i32.pop)
+//     )
+//    )
+//   )
+//
+// Notice that the sequence of instructions in the block is now identical to the
+// sequence of instructions in raw WebAssembly.
+//
+
+#ifndef wasm_ir_stack_h
+#define wasm_ir_stack_h
+
+#include "ir/properties.h"
+#include "wasm-type.h"
+#include "wasm.h"
+
+namespace wasm {
+
+namespace StackUtils {
+
+// Iterate through `block` and remove nops.
+void removeNops(Block* block);
+
+} // namespace StackUtils
+
+// Stack signatures are like regular function signatures, but they are used to
+// represent the stack parameters and results of arbitrary sequences of stacky
+// instructions. They have to record whether they cover an unreachable
+// instruction because their composition takes into account the polymorphic
+// results of unreachable instructions.
+struct StackSignature {
+  Type params;
+  Type results;
+  bool unreachable;
```

Comment:
I'll also add a comment clarifying this in the code, but consider the following sequences of instructions:

```
A:
  i32.add
  unreachable
  f32.const 42

B:
  i32.add
  drop
  f32.const 42
```

The WebAssembly stack type of sequence A is `t1* i32 i32 -> t2* f32` and the WebAssembly stack type of sequence B is `i32 i32 -> f32`. These types are both represented by a `StackSignature` with params `{Type::i32, Type::i32}` and results `Type::f32`, but they need to be distinguished somehow because they are not interchangeable. For example, consider this context:

```
(block (result i64 f32)
 f64.const 0
 i32.const 1
 i32.const 2
 [ ??? ]
)
```

The `[ ??? ]` could be replaced with sequence A, but not with sequence B. The `unreachable` field captures whether the instructions corresponding to a `StackSignature` can be used in contexts that require the type to be polymorphically expanded like this.

Reply:
So it seems `StackSignature`s are computed in two methods: one is `StackSignature` constructor, and the other is `operator+=`. When you use the constructor, there seems to be no way to get a `results` type other than `none` when the expression type is `unreachable`. For example,

```wat
(i32.add
  (i32.add
    (i32.const 0)
    (unreachable)
  )
  (i32.const 3)
)
```
This is a single expression, and its type is `unreachable`, so `unreachable` will be true and its `results` type will be `none`. But if written in the flat IR,
```wat
i32.const 0
unreachable
i32.add
i32.const 3
i32.add
```
According to your explanation, for this code sequence, shouldn't the `results` type be `(t2* i32)`?

Reply:
Also, where do we do such replacements? They are not included in this PR yet I guess?